New module: ezjail - manage FreeBSD jails with ezjail

hc42 opened the pull request:
##### SUMMARY
I created this module to be able to define the state of an FreeBSD jail, managed with Ezjail, in a playbook or role. The module should only do the minimal work of creating or deleting and starting or stopping a jail (similar to the service module). This enables me to quickly setup jails for applications. The installation and configuration of the jailhost should be done by in a playbook since this differs as different people have different preferences and different use cases. This module should be as general as possible.
See https://github.com/hc42/ansible-ezjail for a test playbook. There is also a vagrant file to setup a FreeBSD VM and test the playbook on it.

##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
new module: ezjail (system)

##### ANSIBLE VERSION
```
ansible 2.5.0 (devel 6b14eb2dec) last updated 2018/01/19 21:44:10 (GMT +200)
  config file = None
  configured module search path = ['~/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = ~/ansible-hc/lib/ansible
  executable location = ~/ansible-hc/bin/ansible
  python version = 3.6.4 (default, Jan 12 2018, 19:29:13) [GCC 4.2.1 Compatible Apple LLVM 9.0.0 (clang-900.0.39.2)]
```

Commit:
New module: ezjail - manage FreeBSD jails with ezjail

Commit:
fix: module ezjail - added version

pilou- commented on lib/ansible/modules/system/ezjail.py:
```diff
@@ -0,0 +1,386 @@
+        if running. C(restarted) also creates the jail if not present
+        and starts or restarts it. C(absent) will delete the jail if
+        present (Caution This will also wipe the jailroot. Consider
+        stopping and disbaling the jail if you want to keep the jailroot.)
```
s/disbaling/disabling/

pilou- commented on lib/ansible/modules/system/ezjail.py:
```diff
@@ -0,0 +1,386 @@
+    ''' execute ezjail-admin command '''
+
+    cmd = module.get_bin_path('ezjail-admin', required=True)
+    return module.run_command(' '.join([cmd] + list(params)))
```
[`check_rc=True`](https://github.com/hc42/ansible/blob/6fac750b72d9b4eb6ad12fe9649f8fe8166abc54/lib/ansible/module_utils/basic.py#L2643) parameter of `run_command` could be used instead of raising `FailedException` when `rc != 0`. This will allow to remove caller checks on `rc` and `fail_json` will be called automatically.

pilou- commented on lib/ansible/modules/system/ezjail.py:
```diff
@@ -0,0 +1,386 @@
+    try:
+        conf_file = open(filename, 'r')
+        for line in conf_file:
+            if re.search(
```
Just to let you know, max line length for Ansible project is [160](https://github.com/hc42/ansible/blob/6fac750b72d9b4eb6ad12fe9649f8fe8166abc54/CODING_GUIDELINES.md#pep-8-and-basic-style-checks).

hc42 replied:
Ok, thanks. I don't expect lines much larger as 100. In my tests max. length was about 50.

pilou- commented on lib/ansible/modules/system/ezjail.py:
```diff
@@ -0,0 +1,386 @@
+                line = line.replace(
+                    '"' + old_ip_addr + '"',
+                    '"' + ip_addr + '"')
+            print(line)
```
Debug `print` statement must be removed

hc42 replied:
In this case print is no debug statement but actually writes the content of the file. Removing  it would result in an empty file.

pilou- commented on lib/ansible/modules/system/ezjail.py:
```diff
@@ -0,0 +1,386 @@
+    if new_jail:
+        change_msgs.append('created')
+        if not module.check_mode:
+            if 'ip_addr' not in module.params or module.params['ip_addr'] is None:
```
`'ip_addr' not in module.params` is always `False`

pilou- reviewed (commented):
Did you consider adding integration tests ? A FreeBSD instance is available for integration tests. Here are some integration tests which use FreeBSD:
- https://github.com/ansible/ansible/tree/devel/test/integration/targets/filesystem
- https://github.com/ansible/ansible/blob/devel/test/integration/targets/xml/tasks/main.yml

Commit:
Module Ezjail:
  several bug fixes
  code improvements
  added integration tests

Commit:
Module Ezjail:
  avoid jinja2 filters in test asserts to be more complaint with ansible interation test rules

hc42 commented:
@pilou- I added some tests to cover the basic functionality and most of the possible states. Shippable reports no tests run under the freebsd environments but tests run under the other environment. Not sure if this is the way it should be. I tried to guess as good as possible how the aliases file should look like (due to incomplete documentation of this part of testing and no FreeBSD only examples). Hope this is about what you expected.

Commit:
module ezjail: fix restarted state

Commit:
module ezjail: documentation fixes

Commit:
module ezjail: documentation fixes

Commit:
ezjail: fix test os detection

Commit:
ezjail: fix test os detection

Commit:
module ezjail: integration test guessing (lack of docs)

hc42 commented:
@AugustusKling @ColOfAbRiX @DavidWittman @EvanK @LinusU @abulimov @adejoux @agaffney @ahtik @Akasurde @azaghal @dankeder @davixx @dougluce @dsummersl @giovannisciortino @goozbach @groks @haad @hryamzik @indrajitr @jasperla @jhoekx @jsumners @jtyr @kairoaraujo @kevensen @lberruti @matze @mcv21 @molekuul @mpdehaan @mulby @natefoo @nibalizer @ovcharenko @pilou- @pmarkham @pyykkis @risaacson @rosmo @saito-hideki @sebastiendarocha @sfromm @tdtrask @tmshn @xen0l

I give up on guessing how the tests work and why shippable freebsd always says "WARNING: All targets skipped.". Any help appreciated. But I have no time for guessing and reverse engineering any more.

maxamillion commented:
cc @mattclay - advisement on FreeBSD integration testing?

mattclay commented:
@hc42 The tests appear to be running just fine:

https://app.shippable.com/github/ansible/ansible/runs/56865/51/console

I'm guessing you were looking at the results for the group 1 or 2 tests, but the tests you've added are configured to run in group 3:

https://github.com/ansible/ansible/pull/35111/files#diff-382c42586126ab33afd08a700b0e81d3R1

cinemast commented:
Any updates on this? What is missing to get it merged?

mattclay commented on test/integration/targets/ezjail/aliases:
```diff
@@ -0,0 +1,3 @@
+posix/ci/group3
```
The CI groups have changed since this PR was created:

```suggestion
shippable/posix/group3
```

Akasurde commented:
Please submit a new module request to https://github.com/ansible-collections/community.general. Thanks for your contribution.
